File: content/snippets/go/interview_lcci/16/17_contiguous_sequence.go
package main

import "fmt"

func maxSubArray(nums []int) int {
	best, cur := nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		if cur+nums[i] > nums[i] {
			cur = cur + nums[i]
		} else {
			cur = nums[i]
		}
		if cur > best {
			best = cur
		}
	}
	return best
}

func main() {
	fmt.Println(maxSubArray([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4}))
}

File: content/snippets/cpp/interview_lcci/16/17_contiguous_sequence.cpp
#include <iostream>
#include <vector>
using namespace std;

int maxSubArray(const vector<int>& nums) {
    int best = nums[0];
    int cur = nums[0];
    for (int i = 1; i < static_cast<int>(nums.size()); ++i) {
        cur = max(nums[i], cur + nums[i]);
        best = max(best, cur);
    }
    return best;
}

int main() {
    cout << maxSubArray({-2, 1, -3, 4, -1, 2, 1, -5, 4}) << endl;
    return 0;
}
